examples/multigpu_advtrain/evaluator.py

```diff
@@ -0,0 +1,333 @@
+            self.attacks[att_type] = (col_name, adv_x, preds_adv)
+            # visualize adversarial image
+            tf.summary.image(att_type, adv_x, max_outputs=10)
+        self.sum_op = tf.summary.merge_all()
+
+    def log_value(self, tag, val, desc='', print_stats=False,
+                  print_only=False):
+        if print_stats:
+            valp = ''
+            if not print_only:
+                for i in range(0, 101, 10):
+                    pr = np.percentile(val, i)
+                    self.summary.value.add(tag=tag, simple_value=pr)
+                    valp += ' %d: %.4f' % (i, pr)
+            logging.info('%s (%s):\t %s' % (desc, tag, valp))
+        else:
+            logging.info('%s (%s): %.4f' % (desc, tag, val))
+            if not print_only:
+                self.summary.value.add(tag=tag, simple_value=val)
+
+    def eval_train(self, sess, x, y, preds, X_train, Y_train):
+        subsample_factor = 100
+        X_train_subsampled = X_train[::subsample_factor]
+        Y_train_subsampled = Y_train[::subsample_factor]
+        acc_train = model_eval(sess, x, y, preds, X_train_subsampled,
+                               Y_train_subsampled, args=self.eval_params)
+        self.log_value('train_accuracy_subsampled', acc_train,
+                       'Clean accuracy, subsampled train')
+
+    def eval_test(self, sess, x, y, preds, X_test, Y_test):
+        # Evaluate the accuracy of the MNIST model on legitimate test examples
+        acc = model_eval(sess, x, y, preds, X_test, Y_test,
+                         args=self.eval_params)
+        self.log_value('test_accuracy_natural', acc,
+                       'Clean accuracy, natural test')
+
+    def eval_advs(self, sess, x, y, preds_adv, X_test, Y_test, att_type,
+                  col_name):
+        # Evaluate the accuracy of the MNIST model on adversarial examples
+        end = (len(X_test) // self.batch_size) * self.batch_size
+
+        if self.hparams.fast_tests:
+            end = 10*self.batch_size
+
+        acc = model_eval(sess, x, y, preds_adv, X_test[:end], Y_test[:end],
+                         args=self.eval_params)
+        print_only = False
+        cond = ('FGSM_' in att_type or 'BasicIter_' in att_type
+                or 'PGD_' in att_type)
+        if cond:
+            print_only = True
+        self.log_value('test_accuracy_%s' % att_type, acc,
+                       'Test accuracy on adversarial examples',
+                       print_only=print_only)
+        return acc
+
+    def eval_trans_advs(self, sess, x, y, preds, X_test, Y_test, bbox_np):
+        # Evaluate on the transferred adversarial examples
+        acc_bbox = model_eval(sess, x, y, preds, bbox_np,
+                              Y_test[:len(bbox_np)], args=self.eval_params)
+        self.log_value('test_accuracy_trans_advs', acc_bbox,
+                       'Accuracy on transferred')
+
+    def evaluate(self, inc_epoch=True):
```

Comment:
docstring

The other eval methods are specific about what they're evaluating (`eval_trans_adv`) so I'm not quite sure what to expect when I see a generic `evaluate` method.

Reply:
Fixed in 5f7bc3e and 44356a0.